tools/undelete-blocks/main.go

```diff
@@ -0,0 +1,514 @@
+	}
+
+	return result.ToNames(), nil
+}
+
+func listBlocksForTenant(ctx context.Context, bkt objtools.Bucket, tenantID string) ([]ulid.ULID, error) {
+	result, err := bkt.List(ctx, objtools.ListOptions{
+		Prefix:    tenantID,
+		Versioned: true, // using versioned listing in case a block only has deleted objects
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	result.Objects = nil // avoids bucket index versions and any other object noise
+	names, err := result.ToNamesWithoutPrefix(tenantID)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed while listing for block prefixes")
+	}
+
+	blockIDs := make([]ulid.ULID, 0, len(names))
+
+	for _, name := range names {
+		blockID, err := ulid.Parse(name)
+		if err != nil {
+			// block directories will always be a ulid
+			continue
+		}
+		blockIDs = append(blockIDs, blockID)
+	}
+	return blockIDs, nil
+}
+
+func listGlobalDeleteMarkers(ctx context.Context, bucket objtools.Bucket, tenantID string) (map[ulid.ULID]struct{}, error) {
+	prefix := tenantID + objtools.Delim + block.MarkersPathname
+	// Unlike the other listings in this tool, this does not need to be a versioned listing since only the current state is required
+	listing, err := bucket.List(ctx, objtools.ListOptions{
+		Prefix: prefix,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	m := make(map[ulid.ULID]struct{})
+	markers, err := listing.ToNamesWithoutPrefix(prefix)
+	if err != nil {
+		return nil, err
+	}
+	for _, marker := range markers {
+		if ulid, ok := block.IsDeletionMarkFilename(marker); ok {
+			m[ulid] = struct{}{}
+		}
+	}
+
+	return m, nil
+}
+
+func undeleteBlocks(ctx context.Context, bucket objtools.Bucket, blocks map[string][]ulid.ULID, dryRun bool) {
+	succeeded, notNeeded, failed := 0, 0, 0
+	defer func() {
+		slog.Info("undelete operations summary", "succeeded", succeeded, "notNeeded", notNeeded, "failed", failed, "dryRun", dryRun)
+	}()
+
+	for tenantID, blockIDs := range blocks {
+		tenantLogger := slog.With("tenant", tenantID)
+
+		globalDeleteMarkers, err := listGlobalDeleteMarkers(ctx, bucket, tenantID)
+		if err != nil {
+			tenantLogger.Error("failed to list global markers for tenant, failing all blocks for this tenant", "err", err)
+			failed += len(blockIDs)
+			continue
+		}
+
+		for _, blockID := range blockIDs {
+			logger := tenantLogger.With("block", blockID)
+			if err := ctx.Err(); err != nil {
+				logger.Error("context error", "err", err)
+				return
+			}
+
+			_, globalDeleteMarkerExists := globalDeleteMarkers[blockID]
+
+			undeleted, err := undeleteBlock(ctx, bucket, tenantID, blockID, globalDeleteMarkerExists, logger, dryRun)
+			if err != nil {
+				failed++
+				logger.Error("failed to undelete block", "err", err)
+			} else if undeleted {
+				succeeded++
+				if !dryRun {
+					logger.Info("successfully undeleted block")
+				}
+			} else {
+				notNeeded++
+				logger.Info("block did not need to be undeleted")
+			}
+		}
+	}
+}
+
+type version struct {
```

Comment:
This `version` struct and the `restorableVersion` struct don't add much beyond not leaking the `objtools.ObjectAttributes` type too far into the restore logic.  A single struct could probably achieve that same goal more cleanly.